src/lib.rs

```diff
@@ -39,19 +39,62 @@
 //! # }
 //! ```
 
+#[cfg(feature = "serde")]
+extern crate serde;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
 use std::borrow::Borrow;
 use std::collections::hash_map;
 use std::collections::HashMap;
 use std::fmt::{self, Debug};
 use std::hash::Hash;
 
+#[cfg(feature = "serde")]
+#[derive(Eq, Serialize, Deserialize)]
+#[serde(from = "HashMap<K1, (K2, V)>")]
+pub struct MultiMap<K1, K2, V>
+where
+    K1: Eq + Hash + Clone + Debug,
+    K2: Eq + Hash + Clone + Debug,
+{
+    #[serde(flatten)]
+    value_map: HashMap<K1, (K2, V)>,
+    #[serde(skip)]
+    key_map: HashMap<K2, K1>,
+}
+
+#[cfg(feature = "serde")]
+impl<K1, K2, V> From<HashMap<K1, (K2, V)>> for MultiMap<K1, K2, V>
+where
+    K1: Eq + Hash + Clone + Debug,
+    K2: Eq + Hash + Clone + Debug,
+{
+    fn from(tuple_map: HashMap<K1, (K2, V)>) -> Self {
+        let mut m = MultiMap::with_capacity(tuple_map.len());
+        for (k1, (k2, v)) in tuple_map {
+            m.insert(k1, k2, v);
+        }
+        m
+    }
+}
+
+#[cfg(not(feature = "serde"))]
 #[derive(Eq)]
-pub struct MultiMap<K1: Eq + Hash, K2: Eq + Hash, V> {
+pub struct MultiMap<K1, K2, V>
```

Comment:
Do we need to restrict to `Debug` types here?

Reply:
The compiler seems to only need it here `impl<K1, K2, V> fmt::Debug for MultiMap<K1, K2, V>`
So I remove it everywhere else?